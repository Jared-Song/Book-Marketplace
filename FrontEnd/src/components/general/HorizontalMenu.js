import { Button, Grid } from "@material-ui/core";
import React from "react";
import { makeStyles } from "@material-ui/core/styles";

const useStyles = makeStyles((theme) => ({
  selectedItem: {
    width: "100%",
    backgroundColor: "#EBF0FF",
    borderRadius: 15,
    cursor: 'pointer'
  },
  item: {
    width: "100%",
    cursor: 'pointer'
  },
}));

// menuItem shoud looks like: [{title: String, selected: Boolean!, action: () => void}]
export default function HorizontalMenu({ menuItems }) {
  const classes = useStyles();

  return (
    <Grid
      container
      direction="column"
      justifyContent="flex-start"
      alignItems="flex-start"
      spacing={2}
    >
      {menuItems.map((menuItem) => (
        <Grid
          item
          xs={12}
          key={menuItem.title}
          className={menuItem.selected ? classes.selectedItem : classes.item}
        >
          <Button disabled={menuItem.selected} onClick={menuItem.onClick}>{menuItem.title}</Button>
        </Grid>
      ))}
    </Grid>
  );
}
