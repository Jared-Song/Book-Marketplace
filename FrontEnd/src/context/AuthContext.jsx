import React from "react";
import { useJwt } from "react-jwt";

const AuthContext = React.createContext(null);

export const AuthProvider = ({ token, children, loading }) => {
  const [currentToken, setToken] = React.useState(token);
  const { decodedToken } = useJwt(currentToken);
  const [isLoading, setIsLoading] = React.useState(loading);
  React.useEffect(() => {
    setToken(token);
  }, [token]);
  return (
    <AuthContext.Provider
      value={{ currentUser: decodedToken, setToken, isLoading }}
    >
      {children}
    </AuthContext.Provider>
  );
};

export const useCurrentUser = () => React.useContext(AuthContext);
