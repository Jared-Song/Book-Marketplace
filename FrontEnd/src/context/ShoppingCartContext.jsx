import _, { set } from "lodash";
import React from "react";

const ShoppingCartContext = React.createContext(null);

export const ShoppingCartProvider = ({ children }) => {
  const storageItem = typeof window !== "undefined" ? localStorage.getItem("cart") : undefined;
  const [items, setItems] = React.useState(storageItem ? JSON.parse(storageItem) : []);
  React.useEffect(() => {
    localStorage.setItem("cart", JSON.stringify(items));
  }, [items]);
  const removeAllItems = () => {
    setItems([])
    localStorage.clear()
  }

  const removeShoppingCartItem = (id) => {
    const updatedItems = _.remove(items, (item) => {
      return item.id !== id
    })
    setItems(updatedItems)
  }
  const editShoppingCartItem = (newItem) => {
    const updatedItems = items.map((item) => {
      if (newItem.id === item.id) {
        return newItem; 
      }
      return item
    })
    setItems(updatedItems);
  }
  const addIntoShoppingCart = (newItem) => {
    const itemExist = items.filter((item) => item.id === newItem.id);
    if (itemExist.length > 0) {
      const updatedItems = items.map((item) => {
        if (item.id === newItem.id) {
          return {
            ...item,
            quantity: item.quantity + parseInt(newItem.quantity)
          }
        }
        return item;
      })
      setItems(updatedItems);
    } else {
      setItems([
        ...items,
        {
          ...newItem,
          quantity: parseInt(newItem.quantity),
        },
      ]);
    }
  }
  return (
    <ShoppingCartContext.Provider
      value={{
        cartItems: items,
        addIntoShoppingCart,
        editShoppingCartItem,
        removeShoppingCartItem,
        removeAllItems,
      }}
    >
      {children}
    </ShoppingCartContext.Provider>
  );
}

export const useShoppingCart = () => React.useContext(ShoppingCartContext);