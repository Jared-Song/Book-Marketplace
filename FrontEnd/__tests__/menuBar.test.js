import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";
import HorizontalMenu from "../src/components/general/HorizontalMenu";
import PrimarySearchAppBar from "../src/components/general/MenuBar";
import { AuthProvider } from "../src/context/AuthContext";
import { useRouter } from "next/router";

jest.mock("next/router", () => ({
  useRouter() {
    return {
      route: "/",
      pathname: "",
      query: "keyword=test&filter=test",
      asPath: "",
    };
  },
}));

it("renders Search Bar correctly", () => {
  const tree = renderer
    .create(
      <AuthProvider token="">
        <PrimarySearchAppBar />
      </AuthProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
