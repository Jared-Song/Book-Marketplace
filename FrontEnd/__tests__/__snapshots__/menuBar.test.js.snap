// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Search Bar correctly 1`] = `
<div
  className="makeStyles-grow-1"
>
  <header
    className="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
  >
    <div
      className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-text"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          <h6
            className="MuiTypography-root makeStyles-title-3 MuiTypography-h6 MuiTypography-noWrap"
          >
            BOOKEROO
          </h6>
        </span>
      </button>
      <div
        className="makeStyles-grow-1"
      />
      <div
        className="makeStyles-sectionDesktop-8"
      >
        <div
          className="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center"
        >
          <div
            className="MuiGrid-root MuiGrid-item"
          >
            <div
              className="MuiFormControl-root"
            >
              <div
                className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense MuiInput-marginDense"
                onClick={[Function]}
              >
                <div
                  aria-haspopup="listbox"
                  className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
                  onBlur={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onMouseDown={[Function]}
                  role="button"
                  tabIndex={0}
                >
                  Title
                </div>
                <input
                  aria-hidden={true}
                  className="MuiSelect-nativeInput"
                  onAnimationStart={[Function]}
                  onChange={[Function]}
                  required={false}
                  tabIndex={-1}
                  value="title"
                />
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiSelect-icon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
              </div>
            </div>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item"
          >
            <div
              className="makeStyles-search-4"
            >
              <div
                className="makeStyles-searchIcon-5"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                  />
                </svg>
              </div>
              <div
                className="MuiInputBase-root makeStyles-inputRoot-6"
                onClick={[Function]}
              >
                <input
                  aria-label="search"
                  className="MuiInputBase-input makeStyles-inputInput-7"
                  onAnimationStart={[Function]}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  placeholder="Searchâ€¦"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item"
          >
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-outlined"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiButton-label"
              >
                Search
              </span>
            </button>
          </div>
        </div>
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-text"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            login
          </span>
        </button>
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-text"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            signup
          </span>
        </button>
      </div>
    </div>
  </header>
</div>
`;
